clients/src/main/java/org/apache/kafka/common/record/Record.java

```diff
@@ -63,14 +73,37 @@
     public static final int COMPRESSION_CODEC_MASK = 0x07;
 
     /**
+     * Specify the mask of timestamp type.
+     * 0 for CreateTime, 1 for LogAppendTime.
+     */
+    public static final byte TIMESTAMP_TYPE_MASK = 0x08;
+    public static final int TIMESTAMP_TYPE_ATTRIBUTE_OFFSET = 3;
+
+    /**
      * Compression code for uncompressed records
      */
     public static final int NO_COMPRESSION = 0;
 
+    /**
+     * Timestamp value for records without a timestamp
+     */
+    public static final long NO_TIMESTAMP = -1L;
+
     private final ByteBuffer buffer;
+    private final Long wrapperRecordTimestamp;
+    private final TimestampType timestampTypeToUse;
```

Comment:
Would it be better to rename this to wrapperTimestampType? Also, since the way to interpret the timestamp is a bit subtle, especially with respect to compressed messages, could you document this in a comment?